test/e2e/specs/editor/various/autosave.spec.js

```diff
@@ -0,0 +1,351 @@
+		page,
+		pageUtils,
+	} ) => {
+		const notice = page
+			.getByRole( 'button', { name: 'Dismiss this notice' } )
+			.filter( { hasText: 'Draft saved' } );
+
+		await page.keyboard.press( 'Enter' );
+		await page.keyboard.type( 'before save' );
+		await pageUtils.pressKeys( 'primary+s' );
+		await notice.waitFor();
+		await page.keyboard.type( ' after save' );
+		await notice.click();
+
+		// Trigger local autosave.
+		await page.evaluate( () =>
+			window.wp.data.dispatch( 'core/editor' ).autosave( { local: true } )
+		);
+
+		expect(
+			await page.evaluate( () => window.sessionStorage.length )
+		).toBeGreaterThanOrEqual( 1 );
+
+		await page.route(
+			( url ) =>
+				url.href.includes(
+					`rest_route=${ encodeURIComponent( '/wp/v2/posts/' ) }`
+				),
+			async ( route ) => {
+				await route.abort( 'internetdisconnected' );
+			}
+		);
+		await pageUtils.pressKeys( 'primary+s' );
+
+		await expect(
+			page.locator( '.components-notice__content' )
+		).toContainText( 'Updating failed. You are probably offline.' );
+		expect(
+			await page.evaluate( () => window.sessionStorage.length )
+		).toBe( 1 );
+	} );
+
+	// See https://github.com/WordPress/gutenberg/pull/17501.
+	test( "shouldn't conflict with server-side autosave", async ( {
+		editor,
+		page,
+	} ) => {
+		await page.keyboard.press( 'Enter' );
+		await page.keyboard.type( 'before save' );
+		await editor.publishPost();
+
+		const paragraph = editor.canvas.getByRole( 'document', {
+			name: 'Block: Paragraph',
+		} );
+		await paragraph.click();
+		await page.keyboard.type( ' after save' );
+
+		// Trigger remote autosave.
+		await page.evaluate( () =>
+			window.wp.data.dispatch( 'core/editor' ).autosave()
+		);
+		// Force conflicting local autosave.
+		await page.evaluate( () =>
+			window.wp.data.dispatch( 'core/editor' ).autosave( { local: true } )
+		);
+
+		expect(
+			await page.evaluate( () => window.sessionStorage.length )
+		).toBeGreaterThanOrEqual( 1 );
+
+		await page.reload();
+		await page.waitForFunction( () => window?.wp?.data );
+
+		// FIXME: Occasionally, upon reload, there is no server-provided
+		// autosave value available, despite our having previously explicitly
+		// autosaved. The reasons for this are still unknown. Since this is
+		// unrelated to *local* autosave, until we can understand them, we'll
+		// drop this test's expectations if we don't have an autosave object
+		// available.
+		const stillHasRemoteAutosave = await page.evaluate(
+			() =>
+				window.wp.data.select( 'core/editor' ).getEditorSettings()
+					.autosave
+		);
+		if ( ! stillHasRemoteAutosave ) {
+			return;
+		}
+
+		// Only remote autosave notice should be applied.
+		await expect(
+			page.locator( '.components-notice__content' )
+		).toContainText(
+			'There is an autosave of this post that is more recent than the version below.'
+		);
+	} );
+
+	test.skip( 'should clear sessionStorage upon user logout', async ( {
```

Comment:
The actual logout action is causing the follow-up test to fail.  The core uses a [new browser context hack](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/e2e/specs/cache-control-headers-directives.test.js#L20-L22) to test the logout state. I am not sure if it is suitable here.

cc @swissspidy

Reply:
You should be able to log the user out by clearing the `storageState` for that test only:

```js
test( 'should clear sessionStorage upon user logout', async ( {
  test.use( {
    storageState: {}, // User will be logged out.
  } );

  // ...
```

Reply:
Oh, that might not be applicable here - we need a logged-out state mid-test. 

I think that to make a self-contained logout test, we need a copy of the current `storageState` that we can use (and ideally dispose of after the test).

The current storage state path should be under `process.env.STORAGE_STATE_PATH`:

```js
const tmpStorageStatePath = clone( process.env.STORAGE_STATE_PATH );

test( 'should clear sessionStorage upon user logout', async ( { browser } ) => {
  test.use( {
    storageState: tmpStorageStatePath;
  } );

  // ...
} );
```

This way the logout should only affect the tmp storage state file. Does that make sense?

cc: @kevin940726

Reply:
I thought we could only use this at the test file or `test.describe` level. I'm not sure if I can emulate the logout action with it.

Reply:
I'm not sure if I understand the problem exactly. But could we create a temporary new browser context (and page) and log in "manually"? Then we can follow the normal flow to log out too. I think it might be easier than messing with the storage state?

Reply:
Wouldn't creating new context mean that we can't use the provided fixtures though?

Reply:
Yes, but we can always create new POM from a new page:

```js
const { PageUtils } = require('@wordpress/e2e-test-utils-playwright');

const newPage = await context.newPage();
const pageUtils = new PageUtils( { page: newPage } );
```

Reply:
Not sure here, but we still might need to specify an alternative storage state path in case the default (current) one is used.

Reply:
Okay, I'm still stuck on this one. If any of you fine folks want to give it a try, it would be much appreciated 🙇

Reply:
It's a tough one. I think that we might be facing a design issue here. IMO, logging a user out in one test shouldn't mean it stays logged out for the subsequent tests - we should be able to test a logged-out state whenever necessary. Should we move the login logic from the global setup to the `page` fixture setup, something like in [this worker fixture example](https://playwright.dev/docs/test-fixtures#worker-scoped-fixtures)? I guess it means we'd need to reinitialize the `RequestUtils`, or something to that effect. It might come with some additional time cost, though. What do you think?

Reply:
Sorry for the late reply!

I don't have strong preference about this, and I think we can fix it in a follow-up PR if necessary. Do you think it should be a blocker for this migration PR? I'd prefer to open a new PR exploring other ideas and merge this. Given that it's the last puppeteer PR too! 😆

Reply:
I don't think it's a blocker. I'll open a new issue/PR, and we can continue our explorations there.